### DISCUSSION TITLE
New HIP: Verifiable (Selective) Address Provider (VAP/VSAP)

### DISCUSSION THREAD

**Discussion by befranz** [handshake-org/HIPs#48] [Ideas]
@0xStefan and I have been working on a new HIP and are asking for feedback.

## Abstract

This HIP describes a trustless and verifiable way to get a crypto address associated to a specific Handshake name (TLD).

Like [HIP-2](https://github.com/handshake-org/HIPs/blob/master/HIP-0002.md) an address is provided as a single line, but with an additional signature verifiable by the [signmessagewithname](https://hsd-dev.org/api-docs/#signmessagewithname) Handshake function.

An address is fetched via HTTPS, either
- `static (VAP)`     from any source, in this case the address is the same with every request or
- `selective (VSAP)` from a dedicated service which provides a single selective address with every request

https://github.com/befranz/HIPs/blob/master/HIP-xxx-TCAP.md

**Comment by brandondees** [comment]
nice work. i think that this doesn't need to be delivered over HTTPS strictly -- the upsides of linking the name, address, and signature together in a predictable text format means that this data can be transmitted over any protocol: QR code, sticky note, SMS message, email, roadside billboard, whatever, without losing the authenticity guarantee provided by signature verification.

**Reply by befranz** [reply]
You're absolutely right. The first draft was http/https but then the idea came up to make this as definite as possible. It's still on my list.

**Reply by rithvikvibhu** [reply]
I'm not sure if _any_ means should be allowed. At least needs a bit more thought first.
Not all mediums can update instantly like http(s). You send a sticky note to someone, then you won't have control on when it stops spreading. With http, it immediately updates for all clients requesting new addresses.

And if a name is transferred and finalized (which is the only way to revoke a signature), then it breaks all those signatures that everyone has stored.

Imo latest data must always be accessible (whether over http, ftp, web cam pointed at a billboard). But other means that cannot be updated, shouldn't be used.

**Comment by pinheadmz** [comment]
I think your draft HIP looks really good so far, I have a few ideas but I'll wait until a PR to HIPs is opened? Do you have any lingering questions or discussions around it or are you ready for some feedback?

**Reply by befranz** [reply]
We've got some feedback and are currently building some parts as reference.
Any ideas are welcome, I guess we'll pr this soon.

**Comment by 0xStefan** [comment]
### Here is a VAP provider implementation for anyone who wants to try it
https://www.niami.io/address-alias (You must be logged in)

**Description**
Niami is acting as a VAP Provider that offers users to easily create wallet aliases signed with their name.

Example for a VAP URL is: https://api.niami.io/vap/xn--ss8ht4a/hns

In the TXT record for xn--ss8ht4a (🚚💰) you'll find `TXT VAPURL=https://api.niami.io/vap/xn--ss8ht4a/`

To get the wallet information for xn--ss8ht4a (🚚💰), an app (ex: Bob Wallet) can get the VAP URL from the TXT record and then request https://api.niami.io/vap/xn--ss8ht4a/hns to get the hns address.

The answer will be
`<hns-address> <signature>`

Example:
`hs1q9vq36ef4sw9nma62dv32y3u58vl0vq3lcp49fm Lo/Z2ar+WJMqtPszjovW88gaas5S7d6lOoJgguBiYBUFZZDAogZQqg0H7L/IvOdlBpG3jhIhVhUr3q3wc0mFMw==`

A third party should never trust Niami (or any other VAP provider). Therefore it's important to check if the owner of `xn--ss8ht4a` has signed the address `hs1q9vq36ef4sw9nma62dv32y3u58vl0vq3lcp49fm`.

For Handshake Wallets that is super easy. They just use verifyMessageWithName.
For 3rd party apps we are also working on an easy solution without the need for a node.

**Reply by pinheadmz** [reply]
I would also recommend the signed message include specific reference to VAP, maybe even an exact copy of the TXT that is on chain. Here's why: let's say you and I are chatting offline and signing our chat mesages with our names. You ask me for Falci's donation address and im like oh yeah here it is... now you have some random address signed by me without metadata or explanation. You could use this signed message as a replay attack.

You always have to be very careful when you're signing a message.

Another example consider if I signed the message "Hi it's matthew" with the hns tld `pinheadmz` -- anyone with a copy of that signed message can just replay it to anyone else and effectively pretend it was me.

If the message I signed was `"Hi it's Matthew Zipkin. It's April 12 2022 and I am commenting on https://github.com/handshake-org/HIPs/discussions/48#discussioncomment-2552247"` it would be pretty hard to use that signed message in a replay attack.

So I think part of the verification should be like a closed loop: HNS name points to provider, answer from provider not only points back to HNS name, but points back to the VAP record. That way the verifier is absolutely certain that the address in the signed message is intended by the TLD owner to be their HIP-2 address.

**Reply by befranz** [reply]
Yeah, that's reasonable. Syntax would then be 
`<address> <VAPURL><coin> <signature>`

I've also been thinking about describing an additional multiline (json) option for VAP in this HIP which could look like:
```
<Signature - all following lines as message>
{
"vapurl":"<vapurl><key>",
...
}
```

What do you think?

**Reply by Falci** [reply]
> Syntax would then be
> `<address> <VAPURL><coin> <signature>`

This eliminate the possibility to use a service like sia, where the URL is based on the hash of its content.

**Reply by befranz** [reply]
The proposal is to use a static base link on chain (VAPURL) for all keys/coins you want to set and you don't need to change VAPURL if you changed content. You're right with skynet you can't create the correct file in the first place.

**Reply by 0xStefan** [reply]
> You ask me for Falci's donation address and im like oh yeah here it is... now you have some random address signed by me without metadata or explanation. You could use this signed message as a replay attack.

Some polite pushback here ;) You are totally right when it comes to sign message with address. But we are talking about sign message with name. If you would give Befranz an address that you signed and pretended to be Falci's address. Then Befranz can open Bob, go to Verify Message and fill in the fields:

Name to Verify => falci
Message => The address you provided
Signature => The signature you provided

Result: Invalid Signature

Or in other words: You will never be able to sign a message with a name you don't own.

And there is one very good reason to only use the address in the message: Usability
It's so freaking important that a user can verify the address he is signing. I think we should keep it as simple as possible.

@Falci 
That is true. But VAP is not intended to be used with services like sia. 

HIP-2 => For advanced users who want to be self-sufficient and know how to host files
VAP => For everyday users that want to just have an alias and use a 3rd party to create those aliases (without having to trust them)

**Reply by Falci** [reply]
I guess you misunderstood @pinheadmz's comment.

Imagine that: @pinheadmz and @befranz are chatting in a secure app that signs each individual message.
befranz needs my (Falci) address for some random reason, and since he's talking to pinheadmz, he asks him.
When pinheadmz sends my address to befranz he's not trying to impersonate me, nor still my address.
Totally plausible scenario, right?

But now we have a message with an address (my address) that was signed by pinheadmz. We could pay to that address and say: "that's pinheadmz's address, he signed that!"

**Reply by rithvikvibhu** [reply]
From discord: https://discord.com/channels/822591034202521641/840994858112909403/963472590909890651

We can just prefix with a constant / magic string like:
`HIP-XXXX <hns-address> <signature>`

This way, the signer is aware that it can be used by VAPs.

**Reply by 0xStefan** [reply]
Thanks for explaining @Falci 
Let me go through an example here. So I can follow my own thoughts and come back here in dark time ;)

**Ex: User wants to send 100 HNS to pinheadmz in a Handshake Wallet**
Wallet: Request TXT records for pinheadmz
Wallet: Check if VAPURL is present. Found it!
Wallet: I request VAPURL=https://api.niami.io/vap/pinheadmz/hns
VAP Provider: I give you `hs1qcl3m4hwq2cw6slsml2yqzswgjsxwx7cgpn8jc3 Lo/Z2ar+WJMqtPszjovW...r3q3wc0mFMw==`
Wallet: Got an answer from the provider!
Wallet: Let me check if the signature is valid. I'm calling verifyMessageWithName now
Wallet: Signature is valid!
Wallet: I'm ready to send 100 HNS to hs1qcl3m4hwq2cw6slsml2yqzswgjsxwx7cgpn8jc3

**Potential replay attack if @befranz, @Falci and @0xStefan are working together**
...
Wallet: I request VAPURL=https://api.niami.io/vap/pinheadmz/hns
VAP Provider: I give you `<address-owned-by-falci-signed-with-pinheadmz> A+D6OaTysAD90...TBYWkyGQGF9dOmg==`
Wallet: Got an answer from the provider!
Wallet: Let me check if the signature is valid. I'm calling verifyMessageWithName now
Wallet: Signature is valid! **<= This is very bad!**
...

**Summary**
Alright, if we would work together in the chat-app situation explained by @pinheadmz above, we would be able to impersonate pinheadmz.

**Solution**
Adding one element (besides the address) to the message will solve this. Why not just add HIP-XX at the end of the message, like @rithvikvibhu suggested on Discord? A magic string.
Message to sign/verify: `<address> HIP-XX`

Any objections?

**Reply by 0xStefan** [reply]
> We can just prefix with a constant / magic string like:
> `HIP-XXXX <hns-address> <signature>`

I don't think it hurts to do that. But is it really necessary?
I find the general idea @befranz had so elegant:

`<address> <signature>` or even `<any-data> <signature>`

That's the answer you get when calling a VAPURL received from the root zone or DNS.

So do we really need to add more information here? I don't think so :)

**Reply by Falci** [reply]
The worst case scenario is:
* We are not working together: 
+ Falci: has no idea what is going on here.
+ pinheadmz (the victim): he just sent a message in a chat

If the hypothetical chat happened between **pinheadmz** and you (**0xStefan**) for instance, you could use your host admin power and update the content of `https://api.niami.io/vap/pinheadmz/hns` to this new address.

I'm not saying you would, just that there's a small possibility of shit to happen.

**Reply by 0xStefan** [reply]
Oh right! That would be even easier. 

I think the biggest risk is this:
A bad VAP Provider swaps out the address right before a user signs it. With Niami I implemented a security feature so this cannot happen. As long as a user verifies each address or is only using trusted providers this is not a huge problem. But it's a risk we should keep in mind.

**Reply by pinheadmz** [reply]
Yes I think a magic string or other constant prefix is good, can be defined in the HIP and then doesn't even need to be transmitted in any of the messages, it's just assumed to be prefixed to the address before sigingng / verifying.

To take this one step further though, what if the an exact copy of the TXT was the magic string?

So if my TXT is `"HIPxxxx=https://vsap.sendme/hipxxxx/pinheadmz/"` then the message I sign is that string concatenated to the HNS address provided by the provider. I'm not sure if there is even an attack vector available here, but this is nice double check that the expected provider is providing data explicitly given them to proivde by the end recipient

**Reply by befranz** [reply]
I agree but as magic string I would prefer the URL ({VAPURL}{key}) which provides the actual information for the key (coin). 

This could be beneficial for another reason we haven't talked about yet. Outcome of a discussion with Matt was to better avoid DNS queries for TXT records since it might be difficult to implement in a non-Handshake wallet or app. Therefore VAPURL on-chain is currently the anchor for the signed data.
So how does a random app without blockchain access get the VAPURL? There could be proxies for this, like niami could provide the VAPURL for a name which the app then can even validate since this URL is (hidden) part of the message to verify.

**Reply by befranz** [reply]
Apps/Wallets without access to Handshake node can get VAPURL via DoH (easyhandshake.com) too.

**Reply by Falci** [reply]
>a non-Handshake wallet or app...

... won't be able to validate the signature and will need to trust in these proxies

**Reply by befranz** [reply]
According to @0xStefan verifymessagewithname works without access to Handshake node.

**Reply by befranz** [reply]
Based on the discussion I've created a new, stripped-down HIP document 
https://github.com/befranz/HIPs/blob/master/HIP-xxxx-VAP.md

**Reply by pinheadmz** [reply]
> According to @0xStefan verifymessagewithname works without access to Handshake node.

Well I don't think so, you can verify pubkey/msg/sig sure, but you need at least an urkel proof and the full chain of blockheaders to verify that the pubkey belongs to the name.

**Reply by befranz** [reply]
VAPURL and name address as TXT records are necessary to verify that the signature was created with this address/name.

In case the name is transferred to a new owner, he/she is responsible to change those entries as well as nameserver etc.

**Reply by 0xStefan** [reply]
> According to @0xStefan verifymessagewithname works without access to Handshake node.

It's possible if you can get the owner address from a trusted source. There is still some trust involved that both API providers don't collude. However, if you choose wisely the chance for that can be extremely low.

Example:
A website wants to implement VAP aliases. They don't have a node for whatever reason.
The site is using the Niami API to get the receiving address and the signature for @pinheadmz (Niami is using the VAP Standard).
To verify the signature, the website is using Infura to get the current hns owner address of pinheadmz.
With that information they can verify the message without a node.

Maybe there is a more elegant way to solve that in the future. I just wanted to proof that there is a way with minimal trust to enable VAP without a Handshake node.

**Reply by pinheadmz** [reply]
> Maybe there is a more elegant way to solve that in the future. I just wanted to proof that there is a way with minimal trust to enable VAP without a Handshake node.

Ah this is a good design goal -- but we are assuming the sender has access to HNS name resolution in a secure way? It might actually make the whole thing easier if the VAP TXT also included a public key just for this purpose

**Reply by 0xStefan** [reply]
> we are assuming the sender has access to HNS name resolution in a secure way?

Have we ever discussed that @befranz? I'm also not sure how that could help 

> It might actually make the whole thing easier if the VAP TXT also included a public key just for this purpose

Doesn't this open up a new attack surface? Since the TXT records are most likely set by the VAP provider (ex: Niami), the provider could lie and set a public key he owns. Sure, the user could verify it. But does he?

#### On a more general note
I'm in no way saying that VAP should be used without a Handshake node. The opposite is the case. We shouldn't even advertise this as a feature. Especially in the beginning. I just wanted to make sure there isn't a closed door if this was ever requested. Because as far as my understanding goes, MetaMask and many other crypto wallets are mostly relying on API Endpoints (which we basically all trust). 
In the beginning VAP will be implemented in Bob Wallet, then in the Bob Extension. Bob Extension will probably need a new "Verify VAP Information" function. This new function should be implemented into websites like hnschat. So users can verify themselves if the information on a website is valid. If we have come that far we can discuss further. ;)

**Reply by befranz** [reply]
The topic "verifying signature without access to HNS" caused obviously some confusion. To not mix up things further, these are the assumptions we've made for now:

- VAPURL is set on chain by name owner (and not by VAP Provider!)
- Creating and verifying signatures happen by wallets having access to Handshake node
- Signed data includes VAPURL to avoid replay attacks
- We don't trust the data the wallet fetches from {VAPURL}{Key} is correct
- We basically trust that the wallet uses VAP correctly. In case we're not sure, we can verify data ourselves.

As long as only Handshake wallets can use VAP (because others can't verify securely in this configuration) it's pretty much limited to HNS...

And here came my question "How can non-Handshake wallets/apps verify the signed data without any trust involved?"

1. They need the code for the verify function
2. They need some way to get VAPURL (doesn't have to be secure, it's just a reference to find data)
3. They need a SECURE way to get the address which owns the Handshake name for the verify function

For (3.) I proposed that the name owner set a TXT record on chain like VAPADDR=hs1q.... which any wallet can then fetch together with VAPURL for example via DNS-over-HTTPS (as the most simple way) and secured via DNSSEC. In case a NS is set up, both TXT records (VAPURL and VAPADDR) need to be set on NS too.

My question is now, should we add this option for non-Handshake wallets/apps?

**Reply by pinheadmz** [reply]
For (3) we can try "stateless SPV proofs" but they can be attacked at a certain cost, so I'd suggest not using it for big value transactions.

Client would need:
- 6+ block headers (more headers = more secure assumption)
- urkel proof of namestate in the first header (confirmed by the other headers aka buried under PoW)
- the entire transaction indicated in the `owner` field of the namestate

The tx output has the name's "verify" address. The namestate has the txid and output index (as well as the root zone records including the VAP TXT). The urkel proof authenticates the namestate and the more headers you pile on top and verify the more confident the client can be that these headers actually come from HNS mainnet blockchain

**Reply by befranz** [reply]
> For (3) we can try "stateless SPV proofs"

Having this would be great (thanks for explaining this to me in simple terms off records). Assuming that HIP-2 is as secure as DNS is, the same I guess will apply to VAP with your proposed solution.

For now Stefan and I decided to stick to the assumption that both signer and verifier have access to a Handshake node.

**Reply by befranz** [reply]
If there's no further feedback, can we PR this https://github.com/befranz/HIPs/blob/master/HIP-xxxx-VAP.md ?

If so, do I need to rename it?

**Reply by pinheadmz** [reply]
@Falci will asign you HIP numbers

**Reply by Falci** [reply]
Yes, after we have a PR.

**Comment by Falci** [comment]
I miss nested thread.

@pinheadmz the solution for the replay attach you mentioned should be implemented in the chat app: it should not sign individual message as plain text. It should add a prefix/suffix or a json structure:
```
signwithname pinheamdz JSON.stringify({from: "pinheadmz", to: "befranz", message: "hello world" })
```
